lib/owners.rb

```diff
@@ -1,22 +1,48 @@
+require "owners/parser"
 require "owners/version"
 
 class Owners
-  FORMAT = /
-    ^                   # beginning of line
-    \s*
-    (@?[\w@\.\/-]+)     # username, team, or email address
-    \s*
-    ([\w\*\?\.\s-]+)*   # file globs
-    $                   # end of line
-  /x
-
   def initialize(string)
-    @lines = string.scan(FORMAT).map { |match| match.compact }
+    @parser = Parser.new
+    @tokens = @parser.parse(string)
+    @users, @glob_map = parse_tokens(@tokens)
   end
 
   def for(path)
-    @lines.map do |user, globs|
-      user if !globs || globs.split(" ").any? {|glob| File.fnmatch?(glob, path) }
-    end.compact
+    results = []
+    results.concat(@users)
+
+    @glob_map.each_pair do |glob, users|
+      results.concat(users) if File.fnmatch?(glob, path)
+    end
+
+    results.compact
+  end
+
+  private
+
+  def parse_tokens(tokens)
```

Comment:
I'm not terribly happy with this method, but I think it's good enough for now. My main problem is that after we've already tokenized everything, we're iterating over all the tokens again, just to extract "global owners" versus "owners associated with a glob". In a better world, we could just build these mappings while tokenizing the file directly (the `parse` method in _parser.y_). 

I can make that change if it's desired but I'd rather wait until this becomes a serious need/problem!